lib/robot-controller/verify.rb

```diff
@@ -0,0 +1,209 @@
+# verification class
+module RobotController
+  # Usage:
+  #   RobotController::Verify.new('robot1' => 1, 'robot2' => 2, 'robot3' => 0)
+  #   => {
+  #     'robot1': { state: :up, running: 1 },
+  #     'robot2': { state: :down, running: 0 },
+  #     'robot3': { state: :not_enabled, running: 0 }
+  #   }
+  # ----
+  #
+  # When no errors are detected, the output looks like so:
+  #   % bundle exec controller verify
+  #   OK
+  #
+  #   % bundle exec controller verify --verbose
+  #   OK robot1 is up
+  #   OK robot2 is up
+  #   OK robot3 is not enabled
+  #   OK robot4 is not enabled
+  #
+  # If robot2 were down and robot3 were up, the output would look like so:
+  #
+  #   % bundle exec controller verify
+  #   ERROR robot2 is down (0 out of 3 processes running)
+  #   ERROR robot3 is not enabled but 1 process is running
+  #
+  #   % bundle exec controller verify --verbose
+  #   OK robot1 is up
+  #   ERROR robot2 is down (0 out of 3 processes running)
+  #   ERROR robot3 is not enabled but 1 process is running
+  #   OK robot4 is not enabled
+  #
+  # The various states are determined as follows:
+  #
+  #   If the robot is enabled:
+  #     OK: all N processes are running
+  #     ERROR: not all N processes are running
+  #   If the robot is NOT enabled:
+  #     OK: no processes are running
+  #     ERROR: 1 or more processes are running
+  #   If the robot is unknown by the suite:
+  #     ERROR: always
+  class Verify
+    attr_reader :robots
+
+    # @param [Hash] nprocesses expected number of processes for all robots
+    def initialize(nprocesses)
+      fail ArgumentError if nprocesses.nil? || !nprocesses.is_a?(Hash)
+      fail ArgumentError, 'Empty argument' if nprocesses.size == 0
+      @running = nprocesses
+      @robots = @running.each_key.to_a
+      @status = nil
+    end
+
+    # @param [Boolean] reload forces a reload of status information
+    # @return [Hash<Hash>] status of all robots
+    # {
+    #   'robot1' : {
+    #     state: :up,
+    #     running: 2
+    #   },
+    #   'robot2' : {
+    #     state: :down,
+    #     running: 0
+    #   },
+    #   'robot3' : {
+    #     state: :not_enabled,
+    #     running: 0
+    #   }
+    # }
+    def verify(reload = true)
+      @status = nil if reload
+      r = {}
+      robots.each do |robot|
+        r[robot] = robot_status(robot)
+      end
+      r
+    end
```

Comment:
Consider a shorter construct to get the `r` hash, like so:
```ruby
robots = [*1..10]
#=> [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
r = Hash[ robots.collect {|r| [r, 'x']} ]
#=> {1=>"x", 2=>"x", 3=>"x", 4=>"x", 5=>"x", 6=>"x", 7=>"x", 8=>"x", 9=>"x", 10=>"x"}
```
In this case it would be:
```ruby
Hash[  robots.collect {|r| [r, robot_status(r)] } ]
```
There is no need to init `r` in this form, the Hash init should return it.  So, 5 lines of code could be a 1 liner.